include subdir support based on config key :subdirs?

chadhs opened the pull request:
Here's a quick initial stab at adding this for https://github.com/cryogen-project/cryogen-core/issues/107

I haven't figured out how to install this locally to test with my blog yet.

Let me know if you think I'm on the write track @lacarmen or @yogthos

lacarmen commented:
I think that looks right so far, you'll have to test to find out :) You can test locally by running `lein install` in the root folder if the project. 

Minor comment: I think the key should be a little more descriptive - perhaps `:collapse-subdirs?`

chadhs commented:
@lacarmen that key name is way better so I changed it to `:collapse-subdirs?`.  i also installed locally, tested, fixed an issue, and everything works locally with my blog.  👍🏻

any refactorings / updates you'd recommend, or do you think this is good?

chadhs commented:
also… i did push to github and all still working in "production" http://www.chadstovern.com ; looks like i'm out of excuses to not start writing again.  😆

madstap commented:
Two questions:

* Should this be a required key? (It would make it nicer to upgrade cryogen versions if it was optional.)
* What should be the default value?

You should also make an entry for `:collapse-subdirs?` in the [docs](https://github.com/cryogen-project/cryogen-docs/blob/master/resources/templates/md/docs/configuration.md).

Let us know if you run into any difficulties.

chadhs commented:
@madstap making this optional would be nicer and i think the default value should be false/off as the option is about collapsing the subdirs and there was a previous change to honor subdirs that is now default behavior.

I'll make this change and provide a PR to update the docs accordingly.

Thank you for the feedback; I really appreciate it.

lacarmen commented:
Good feedback @madstap and thanks for the changes @chadhs :)

chadhs commented:
@lacarmen @madstap i made the minor update above and opened a PR here for the docs update https://github.com/cryogen-project/cryogen-docs/pull/21

let me know if there's any other changes, tweaks, etc on either PR.  thank you.

lacarmen commented:
@chadhs This breaks when post dates are stored in the file name rather than in the page-meta. 

I tested with a fresh Cryogen template and moved the 2014 posts into a `2014` folder and the 2016 post into a `2016` folder. Everything else was unchanged from the default template.

```
carmen@ouroboros [~/s/test] ⋊> lein ring server
loading module: Markdown parser
compiling assets...
                              java.lang.Thread.run              Thread.java:  745
java.util.concurrent.ThreadPoolExecutor$Worker.run  ThreadPoolExecutor.java:  617
 java.util.concurrent.ThreadPoolExecutor.runWorker  ThreadPoolExecutor.java: 1142
               java.util.concurrent.FutureTask.run          FutureTask.java:  266
                                               ...
               clojure.core/binding-conveyor-fn/fn                 core.clj: 1938
                           clojure.core/pmap/fn/fn                 core.clj: 6736
            cryogen-core.compiler/read-posts/fn/fn             compiler.clj:  151
                  cryogen-core.compiler/parse-post             compiler.clj:  131
             cryogen-core.compiler/parse-post-date             compiler.clj:   66
                        java.text.DateFormat.parse          DateFormat.java:  366
               java.text.ParseException: Unparseable date: "2014/2014-"
    errorOffset: 4
java.util.concurrent.ExecutionException: java.text.ParseException: Unparseable date: "2014/2014-"
```
This is because the `file-name` returned by `page-content` is `2014/2014-03-10-first-post.md` and `parse-post-date` (used in `parse-post`) takes the first 10 characters of the file name to parse.

lacarmen commented:
@chadhs ping 😄

chadhs commented:
sorry @lacarmen i'll do my best to sort this out in the next week.

lacarmen commented:
@chadhs no worries, just checking if you're still working on this :)

chadhs commented:
@lacarmen may need to wait until next week; apologies.

chadhs commented:
can confirm this isn't an issue when `collapse-subdirs?` is set to true, will test with set to false / default settings as you did now.

chadhs commented:
@lacarmen i did some more testing and discovered that this bug occurs currently with the current cryogen-core.  

steps to recreate:
- remove cryogen deps from my local .m2 folder
- create a new blog per the docs
- run lein deps and then lein ring server to see that cryogen-core was pulled in and then run the server
- create a 2014 dir
- move a 2014 test post into the 2014 dir → **yields the error you reported**
- add `:data` meta data to the post → error no longer happens, compiles correctly
- add `:collapse-subdirs? true` to config and ensure it does nothing to ensure i'm not somehow being fooled by some cache somewhere and that we're really running current prod cryogen-core 👍🏻

**recommendation**
- since we're not introducing a bug with this PR: merge this PR, and the corresponding doc PR
- open a new issue to work on gracefully handling the scenario which causes this issue.

Let me know your thoughts.  Thanks!

lacarmen commented:
Thanks for investigating @chadhs. I agree with your recommendation, let's open a new issue to address the bug :)

lacarmen commented:
Added #109 for the bug.
